e2e/stack.go

```diff
@@ -62,18 +67,40 @@ type L1User struct {
 	PrivateKey *ecdsa.PrivateKey
 }
 
-// New assumes all ports are available and that all paths exist and are valid.
-func New(
-	anvilURL,
-	monomerEngineURL,
-	monomerCometURL,
-	opNodeURL *url.URL,
-	deployConfigDir string,
-	l1stateDumpDir string,
+// Setup creates and runs a new stack for end-to-end testing.
+//
+// It assumes availability of hard-coded local URLs for the Monomer engine, Comet, and OP node.
+//
+// It returns a StackConfig with L1 and L2 clients, the rollup config, and operator and user accounts.
+func Setup(
+	ctx context.Context,
+	env *environment.Env,
 	prometheusCfg *config.InstrumentationConfig,
 	eventListener EventListener,
-) *Stack {
-	return &Stack{
+) (*StackConfig, error) {
+	deployConfigDir, err := filepath.Abs("./optimism/packages/contracts-bedrock/deploy-config")
+	if err != nil {
+		return nil, fmt.Errorf("abs deploy config dir: %v", err)
+	}
+	l1stateDumpDir, err := filepath.Abs("./optimism/.devnet")
+	if err != nil {
+		return nil, fmt.Errorf("abs l1 state dump dir: %v", err)
+	}
+
+	monomerEngineURL, err := newURL("ws://127.0.0.1:8889")
+	if err != nil {
+		return nil, fmt.Errorf("new l1 url: %v", err)
+	}
+	monomerCometURL, err := newURL("http://127.0.0.1:8890")
+	if err != nil {
+		return nil, fmt.Errorf("new l1 url: %v", err)
+	}
+	opNodeURL, err := newURL("http://127.0.0.1:8891")
+	if err != nil {
+		return nil, fmt.Errorf("new l1 url: %v", err)
+	}
```

Comment:
All of the error messages are the same.
Can the port in each address be `0`?

Reply:
Error messages fixed here: 35bf818

The `0` port idea being that each service would look up and assume some available port at runtime?

Reply:
> The 0 port idea being that each service would look up and assume some available port at runtime?

kernel automatically assigns unused port when `0` is provided

Reply:
I had tried using `:0` initially, and the test hangs.

The reason is that (some of?) the values `monomerEngineURL`, `monomerCometURL`, and `opNodeURL` are read more than once - first to initialize a listener, then later to connect a client to that address.

If we'd like to dynamically assign the cleanest thing is probably to remove these values altogether, and force correct piping of address data down the line.

Reply:
Ahhh that's what I was forgetting. Disregard me then. This is why it's nice when APIs accept `Listener`s so we can learn the address that the kernel assigns.